File: hellojava/src/main/java/com/fathzer/chess/Piece.java
package com.fathzer.chess;

import java.util.HashMap;
import java.util.Map;

public enum Piece {
	BLOCKER('X', true),
	WHITE_PAWN('P', true),
	WHITE_KNIGHT('N', true),
	WHITE_BISHOP('B', true),
	WHITE_ROOK('R', true),
	WHITE_QUEEN('Q', true),
	WHITE_KING('K', true),
	BLACK_PAWN('p', false),
	BLACK_KNIGHT('n', false),
	BLACK_BISHOP('b', false),
	BLACK_ROOK('r', false),
	BLACK_QUEEN('q', false),
	BLACK_KING('k', false);

	private static final Map<Character, Piece> CODE_TO_PIECE = new HashMap<>();
	static {
		for (Piece piece : values()) {
			if (piece != BLOCKER) {
				CODE_TO_PIECE.put(piece.code, piece);
			}
		}
	}

	private final char code;
	private final boolean isWhite;

	private Piece(char code, boolean isWhite) {
		this.code = code;
		this.isWhite = isWhite;
	}

	public static Piece fromCode(char code) {
		return CODE_TO_PIECE.get(code);
	}

	public char getCode() {
		return code;
	}

	public boolean isWhite() {
		return isWhite;
	}

    public boolean canBeCapturedBy(boolean white) {
        return this!=Piece.BLOCKER && white != isWhite;
    }
}

File: hellogo/pkg/chess/Piece.go
package chess

// Capturable interface from MoveBuilder.go
// ...existing code...

type Piece struct {
	Code    rune
	IsWhite bool
	ordinal int
}

// Equals checks if two pieces are equal.
func (p *Piece) Equals(other *Piece) bool {
	if p == other {
		return true
	}
	if p == nil || other == nil {
		return false
	}
	return p.Code == other.Code && p.IsWhite == other.IsWhite
}

var (
	BLOCKER      = Piece{'X', true, 0}
	WHITE_PAWN   = Piece{'P', true, 1}
	WHITE_KNIGHT = Piece{'N', true, 2}
	WHITE_BISHOP = Piece{'B', true, 3}
	WHITE_ROOK   = Piece{'R', true, 4}
	WHITE_QUEEN  = Piece{'Q', true, 5}
	WHITE_KING   = Piece{'K', true, 6}
	BLACK_PAWN   = Piece{'p', false, 7}
	BLACK_KNIGHT = Piece{'n', false, 8}
	BLACK_BISHOP = Piece{'b', false, 9}
	BLACK_ROOK   = Piece{'r', false, 10}
	BLACK_QUEEN  = Piece{'q', false, 11}
	BLACK_KING   = Piece{'k', false, 12}
)

var (
	allPieces = []Piece{
		BLOCKER,
		WHITE_PAWN, WHITE_KNIGHT, WHITE_BISHOP, WHITE_ROOK, WHITE_QUEEN, WHITE_KING,
		BLACK_PAWN, BLACK_KNIGHT, BLACK_BISHOP, BLACK_ROOK, BLACK_QUEEN, BLACK_KING,
	}
	codeToPiece map[rune]*Piece
)

func init() {
	codeToPiece = make(map[rune]*Piece, len(allPieces))
	for i := range allPieces {
		p := &allPieces[i]
		if p.Code != BLOCKER.Code {
			codeToPiece[p.Code] = p
		}
	}
}

// FromCode returns a pointer to the Piece for a given character code, or nil if not found.
func FromCode(code rune) *Piece {
	if p, ok := codeToPiece[code]; ok {
		return p
	}
	return nil
}

// GetCode returns the character code for the piece.
func (p *Piece) GetCode() rune {
	return p.Code
}

// IsWhitePiece returns true if the piece is white.
func (p *Piece) IsWhitePiece() bool {
	return p.IsWhite
}

// CanBeCapturedBy implements the Capturable interface.
func (p *Piece) CanBeCapturedBy(white bool) bool {
	return p != nil && p != &BLOCKER && white != p.IsWhite
}

// Ordinal returns the ordinal value of the piece, similar to Java's Enum.ordinal().
func (p *Piece) Ordinal() int {
	return p.ordinal
}
